### PR TITLE
26) passed all tasks by Dao8

### PR DESCRIPTION
**Pull request by AntonPanchishko**

### CODE REVIEW

File: src/main/java/taxi/controllers/car/AddCarController.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import taxi.lib.Injector;
+import taxi.model.Car;
+import taxi.model.Manufacturer;
+import taxi.service.CarService;
+import taxi.service.ManufacturerService;
+
+public class AddCarController extends HttpServlet {
+    private static final Injector injector = Injector.getInstance("taxi");
```
Comment (Max-Pochepets) [inline src/main/java/taxi/controllers/car/AddCarController.java]:
```suggestion
    private static final Injector INJECTOR = Injector.getInstance("taxi");
```
https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html#:~:text=Class%20names%20should%20be%20nouns,such%20as%20URL%20or%20HTML

File: src/main/java/taxi/controllers/car/CarController.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.List;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import taxi.lib.Injector;
+import taxi.model.Car;
+import taxi.service.CarService;
+
+public class CarController extends HttpServlet {
```
Comment (Max-Pochepets) [inline src/main/java/taxi/controllers/car/CarController.java]:
```suggestion
public class GetAllCarController extends HttpServlet {
```

File: src/main/java/taxi/controllers/car/DeleteCarController.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    private final CarService carService = (CarService)
+            injector.getInstance(CarService.class);
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        String carId = req.getParameter("car_id");
+        Long id = Long.valueOf(carId);
+        carService.delete(id);
+        resp.sendRedirect(req.getContextPath() + "/cars/all");
```
Comment (Max-Pochepets) [inline src/main/java/taxi/controllers/car/DeleteCarController.java:22]:
```suggestion
        resp.sendRedirect(req.getContextPath() + "/cars/");
```
just a suggestion

File: src/main/webapp/WEB-INF/views/cars/add.jsp [hunk 4]
```diff
@@ -0,0 +1,14 @@
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<html>
+<head>
+    <title>Add car</title>
+</head>
+<body>
+<h1>Provide new car</h1>
+<form method="post" action="${pageContext.request.contextPath}/cars/add">
+    Enter manufacturer_id<input type="text" name="manufacturer_id"><br>
```
Comment (Max-Pochepets) [inline src/main/webapp/WEB-INF/views/cars/add.jsp]: 1. what if manufacturer with input id doesn't exist?
2. what if you input "2dasf234dafa" in this field?
3. what if you don't input anything in this field?

File: src/main/webapp/web.xml [hunk 5]
```diff
@@ -9,6 +9,96 @@
     </servlet>
     <servlet-mapping>
         <servlet-name>index</servlet-name>
-        <url-pattern>/index</url-pattern>
+        <url-pattern>/</url-pattern>
```
Comment (Max-Pochepets) [inline src/main/webapp/web.xml:12]: where do you call your controllers? I didn't see you modifying index controller or jsp. what does your home page show?

**Review by Max-Pochepets — CHANGES_REQUESTED** [review]

File: src/main/webapp/web.xml (hunk 5 above)
Comment (dmytro-hryhoruk) [reply src/main/webapp/web.xml:12]: home page or some kind of menu is not required in the scope of this task

File: src/main/java/taxi/controllers/driver/DriversController.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import java.util.List;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import taxi.lib.Injector;
+import taxi.model.Driver;
+import taxi.service.DriverService;
+
+public class DriversController extends HttpServlet {
```
Comment (dmytro-hryhoruk) [inline src/main/java/taxi/controllers/driver/DriversController.java]:
```suggestion
public class GetAllDriversController extends HttpServlet {
```

File: src/main/webapp/web.xml (hunk 5 above)
Comment (Max-Pochepets) [reply src/main/webapp/web.xml:12]: > home page or some kind of menu is not required in the scope of this task

fair point, but how should we test our application? =)

File: src/main/java/taxi/controllers/manufacturer/ManufacturerController.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.List;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import taxi.lib.Injector;
+import taxi.model.Manufacturer;
+import taxi.service.ManufacturerService;
+
+public class ManufacturerController extends HttpServlet {
```
Comment (dmytro-hryhoruk) [inline src/main/java/taxi/controllers/manufacturer/ManufacturerController.java]:
```suggestion
public class GetAllManufacturersController extends HttpServlet {
```

File: src/main/webapp/WEB-INF/views/cars/add.jsp (hunk 4 above)
Comment (dmytro-hryhoruk) [reply src/main/webapp/WEB-INF/views/cars/add.jsp]: @AntonPanchishko about @Max-Pochepets comment: 
1) you may handle this in your controller and redirect to some error page in this case(but it is not required in the scope of this task);
2) You may Handle NumberFormatException and show some message near input("invalid input etc. try again)
3) You may use required attribute of input tag;

File: src/main/webapp/WEB-INF/views/cars/drivers/add.jsp [hunk 8]
```diff
@@ -0,0 +1,14 @@
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<html>
+<head>
+    <title>Add driver to car</title>
+</head>
+<body>
+<h1>Provide driver to car</h1>
+<form method="post" action="${pageContext.request.contextPath}/cars/drivers/add">
+    car id: <input type="number" name="car_id" required><br>
+    driver id: <input type="number" name="driver_id" required><br>
```
Comment (dmytro-hryhoruk) [inline src/main/webapp/WEB-INF/views/cars/drivers/add.jsp]:
```suggestion
    car id: 
    <input type="number" name="car_id" required><br>
    driver id: 
    <input type="number" name="driver_id" required><br>
```

**Review by dmytro-hryhoruk — CHANGES_REQUESTED** [review]
Good job!

**Review by Max-Pochepets — APPROVED** [review]

File: src/main/java/taxi/controllers/car/AddCarController.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        try {
+            String model = req.getParameter("model");
+            Long manufacturerId = Long.parseLong(req.getParameter("manufacturer_id"));
+            Manufacturer manufacturer = manufacturerService.get(manufacturerId);
+            Car car = new Car(model, manufacturer);
+            carService.create(car);
+            resp.sendRedirect(req.getContextPath() + "/");
+        } catch (NoSuchElementException e) {
+            req.setAttribute("message", "You put incorrect data. No manufacturer with such ID");
+            req.getRequestDispatcher("/WEB-INF/views/cars/add.jsp").forward(req, resp);
```
Comment (dmytro-hryhoruk) [inline src/main/java/taxi/controllers/car/AddCarController.java:40]: let's not catch the runtime exception. Int the scope of this task we can just believe that we will get the proper id(remove this in each controller). Add all validation on jsp pages such as input type and required attributes.

File: src/main/java/taxi/controllers/car/DeleteCarController.java (hunk 3 above)
Comment (dmytro-hryhoruk) [reply src/main/java/taxi/controllers/car/DeleteCarController.java:22]: let's fix this then

File: src/main/webapp/web.xml [hunk 10]
```diff
@@ -9,6 +9,98 @@
...
+        <url-pattern>/manufacturers/delete</url-pattern>
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>allCars</servlet-name>
+        <servlet-class>taxi.controllers.car.GetAllCarController</servlet-class>
+    </servlet>
+    <servlet-mapping>
+        <servlet-name>allCars</servlet-name>
+        <url-pattern>/cars/all</url-pattern>
```
Comment (dmytro-hryhoruk) [inline src/main/webapp/web.xml:93]:
```suggestion
        <url-pattern>/cars</url-pattern>
```

File: src/main/webapp/web.xml [hunk 11]
```diff
@@ -9,6 +9,98 @@
...
+        <url-pattern>/cars/drivers/add</url-pattern>
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>allManufacturers</servlet-name>
+        <servlet-class>taxi.controllers.manufacturer.GetAllManufacturerController</servlet-class>
+    </servlet>
+    <servlet-mapping>
+        <servlet-name>allManufacturers</servlet-name>
+        <url-pattern>/manufacturers/all</url-pattern>
```
Comment (dmytro-hryhoruk) [inline src/main/webapp/web.xml:66]:
```suggestion
        <url-pattern>/manufacturers</url-pattern>
```

File: src/main/webapp/web.xml (hunk 5 above)
Comment (dmytro-hryhoruk) [reply src/main/webapp/web.xml:12]: @Max-Pochepets manually typing the url's in the search bar

File: src/main/webapp/web.xml [hunk 12]
```diff
@@ -9,6 +9,98 @@
...
+        <url-pattern>/</url-pattern>
     </servlet-mapping>
+
+    <servlet>
+        <servlet-name>allDrivers</servlet-name>
+        <servlet-class>taxi.controllers.driver.GetAllDriversController</servlet-class>
+    </servlet>
+    <servlet-mapping>
+        <servlet-name>allDrivers</servlet-name>
+        <url-pattern>/drivers/all</url-pattern>
```
Comment (dmytro-hryhoruk) [inline src/main/webapp/web.xml:21]:
```suggestion
        <url-pattern>/drivers</url-pattern>
```

**Review by dmytro-hryhoruk — CHANGES_REQUESTED** [review]

**Pull request closed without merging**